L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+    outCollection.reset(new l1t::HGCalTriggerCellBxCollection);
+
+  int bx(triggerCells.getFirstBX());
+  unsigned bxOffset(0);
+
+  typedef edm::Ptr<l1t::HGCalTriggerCell> TriggerCellPtr;
+  typedef edm::Ptr<l1t::HGCalCluster> ClusterPtr;
+
+  // ClusteringDummyImpl only considers BX 0, so we dump all cells to one vector
+  std::vector<TriggerCellPtr> triggerCellPtrs;
+
+  // loop through all bunch crossings
+  for (unsigned iC(0); iC != triggerCells.size(); ++iC) {
+    if (iC >= bxOffset + triggerCells.size(bx)) {
+      bxOffset += triggerCells.size(bx);
+      bx += 1;
+    }
+
+    auto& cell(triggerCells[iC]);
+
+    auto mapElem(tcToCalo.find(cell.detId()));
+    if (mapElem == tcToCalo.end())
+      continue;
+
+    edm::Ref<l1t::HGCalTriggerCellBxCollection> ref(triggerCellsHandle, iC);
+
+    outMap->insert(mapElem->second, ref);
+
+    if (makeCellsCollection_)
+      outCollection->push_back(bx, cell);
+
+    triggerCellPtrs.emplace_back(triggerCellsHandle, iC);
+  }
+
+  _event.put(std::move(outMap));
+  if (makeCellsCollection_)
+    _event.put(std::move(outCollection), "HGCalVFEProcessorSums");
+
+  auto outClusters(std::make_unique<l1t::HGCalClusterBxCollection>());
+
+  auto sortCellPtrs([](TriggerCellPtr const& lhs, TriggerCellPtr const& rhs)->bool {
+      return lhs->mipPt() > rhs->mipPt();
+    });
+
+  std::sort(triggerCellPtrs.begin(), triggerCellPtrs.end(), sortCellPtrs);
+  dummyClustering_.clusterizeDummy(triggerCellPtrs, *outClusters);
+
+  std::map<CaloParticleRef, std::vector<unsigned>> caloToClusterIndices;
+  for (unsigned iC(0); iC != outClusters->size(); ++iC) {
+    auto& cluster((*outClusters)[iC]);
+    // cluster detId and cell detId are identical
+    auto caloRef(tcToCalo.at(cluster.detId()));
+    caloToClusterIndices[caloRef].push_back(iC);
+  }
+
+  auto&& clustersHandle(_event.put(std::move(outClusters), "HGCalBackendLayer1Processor2DClustering"));
+
+  auto outMulticlusters(std::make_unique<l1t::HGCalMulticlusterBxCollection>());
+
+  for (auto& ref : orderedCaloRefs) {
+    if (ref.isNull()) // shouldn't happen
+      continue;
+
+    auto& caloParticle(*ref);
+
+    l1t::HGCalTriggerCell dummyCell;
+    dummyCell.setDetId(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    TriggerCellPtr tcptr(&dummyCell, 0);
+
+    l1t::HGCalCluster dummyCluster(tcptr);
+    ClusterPtr clptr(&dummyCluster, 0);
+
+    l1t::HGCalMulticluster multicluster(clptr); // set the detid
+    
+    for (unsigned iC : caloToClusterIndices[ref]) {
+      ClusterPtr clPtr(clustersHandle, iC);
+      multicluster.addConstituent(clPtr, true, 1.);
+    }
+
+    multicluster.removeConstituent(clptr); // then remove the dummy
+
+    double sumPt(multicluster.sumPt());
+    auto&& centre(multicluster.centre());
+    math::PtEtaPhiMLorentzVector multiclusterP4(sumPt, centre.eta(), centre.phi(), 0.);
+    multicluster.setP4(multiclusterP4);
+
+    multicluster.showerLength(showerShape_.showerLength(multicluster));
```

Comment:
There is now a `fillShapes` method that can be used, e.g. like [here](https://github.com/PFCal-dev/cmssw/blob/hgc-tpg-devel-CMSSW_10_6_0_pre2/L1Trigger/L1THGCal/src/backend/HGCalHistoClusteringImpl.cc#L162)

Reply:
This is awesome.